File: Source/Foreign/ImageLoader.hs
module Foreign.ImageLoader (loadImage, isImageLoaded, getLoadedImage) where

    import Haste
    import Haste.DOM as DOM
    import Haste.Graphics.Canvas as Canvas
    import Haste.Foreign as Foreign

    loadImage :: String -> IO ()
    loadImage = Foreign.ffi $ Haste.toJSString "IMAGE_LOADER['loadImage']"

    isImageLoaded :: String -> IO Bool
    isImageLoaded = Foreign.ffi $ Haste.toJSString "IMAGE_LOADER['isImageLoaded']"

    getLoadedImageElement :: String -> IO DOM.Elem
    getLoadedImageElement = Foreign.ffi $ Haste.toJSString "IMAGE_LOADER['getLoadedImage']"

    getLoadedImage :: String -> IO Canvas.Bitmap
    getLoadedImage path = do
        element <- getLoadedImageElement path
        bitmap <- Canvas.loadBitmap element
        return bitmap

File: Source/Foreign/ImageLoader.js
(function() {
    var loadedImages = {};
    var hadErrorOccured = false;

    window["IMAGE_LOADER"] = {
        "loadImage" : function (path) {
            if (!(path in loadedImages)) {
                loadedImages[path] = null;

                var imageElement = new Image();

                imageElement.onload = function() {
                    loadedImages[path] = imageElement;
                };

                imageElement.onerror = function(error) {
                    console.error(error);
                    hadErrorOccured = true;
                };

                imageElement.src = path;
            }
        },

        "isImageLoaded" : function(path) {
            return path in loadedImages && loadedImages[path] !== null;
        },

        "getLoadedImage" : function(path) {
            return loadedImages[path];
        },

        "hadErrorOccured" : function() {
            return hadErrorOccured;
        }
    };
})();